subprojects/core/src/integTest/groovy/org/gradle/api/AbstractDomainObjectContainerIntegrationTest.groovy

```diff
@@ -64,6 +64,107 @@ abstract class AbstractDomainObjectContainerIntegrationTest extends AbstractInte
+            "${containerType}#whenObjectAdded(Action)": "testContainer.whenObjectAdded {}",
+            "${containerType}#withType(Class, Action)": "testContainer.withType(testContainer.type) {}",
+            "${containerType}#all(Action)": "testContainer.all {}",
+            "Project#afterEvaluate(Closure)": "afterEvaluate {}",
+            "Project#beforeEvaluate(Closure)": "beforeEvaluate {}",
+            "Gradle#beforeProject(Closure)": "gradle.beforeProject {}",
+            "Gradle#afterProject(Closure)": "gradle.afterProject {}",
+            "Gradle#projectsLoaded(Closure)": "gradle.projectsLoaded {}",
+            "Gradle#projectsEvaluated(Closure)": "gradle.projectsEvaluated {}",
+        ]
+    }
+}
+
+abstract class AbstractQueryAndMutationDomainObjectContainerIntegrationTest extends AbstractDomainObjectContainerIntegrationTest {
+    @Unroll
+    def "can execute query and mutating methods #method.key from all"() {
+        buildFile << """
+            testContainer.all {
+                if (it.name == "realized") {
+                    ${method.value}
+                }
+            }
+        """
+
+        expect:
+        succeeds "help"
+
+        where:
+        method << getQueryMethods() + getMutationMethods()
+    }
+
+    @Unroll
+    def "can execute query and mutating methods #method.key from withType.all"() {
+        buildFile << """
+            testContainer.withType(testContainer.type).all {
+                if (it.name == "realized") {
+                    ${method.value}
+                }
+            }
+        """
+
+        expect:
+        succeeds "help"
+
+        where:
+        method << getQueryMethods() + getMutationMethods()
+    }
+
+    @Unroll
+    def "can execute query and mutation methods #method.key from getByName"() {
+        buildFile << """
+            testContainer.getByName("realized") {
+                ${method.value}
+            }
+        """
+
+        expect:
+        succeeds "help"
+
+        where:
+        method << getQueryMethods() + getMutationMethods()
+    }
+
+    @Unroll
+    def "can execute query and mutation methods #method.key from withType.getByName"() {
+        buildFile << """
+            testContainer.withType(testContainer.type).getByName("realized") {
+                ${method.value}
+            }
+        """
+
+        expect:
+        succeeds "help"
+
+        where:
+        method << getQueryMethods() + getMutationMethods()
+    }
+
+    @Unroll
+    def "can execute query and mutating methods #method.key from create(String)"() {
+        buildFile << """
+            testContainer.create("a") {
+                ${method.value}
+            }
+        """
+        expect:
+        succeeds "help"
+        where:
+        method << getQueryMethods() + getMutationMethods()
+    }
+}
+
+abstract class AbstractQueryDomainObjectContainerIntegrationTest extends AbstractDomainObjectContainerIntegrationTest {
```

Comment:
Can you move these classes to top-level files?

Reply:
Done.